zmq-async/src/deferred.ml

```diff
@@ -35,7 +35,7 @@ module Fd = struct
   type 'a t' = 'a t
   type t = Fd.t
   let create fd =
-    Fd.create (Fd.Kind.Socket `Active) fd (Base.Info.of_string "<zmq>")
+    Fd.create Fd.Kind.File fd (Base.Info.of_string "<zmq>")
```

Comment:
So if I understand correctly, an `Fd.t` is created and the error is that here we specify a socket but in reality it is a `file` and this difference is only relevant on `s390x` because presumably other platforms don't particularly care?

My second comment was exactly "can we determine this automatically instead of hardcoding" which is preempted by the fact that this is an API change. Annoying, but I understand the rationale. An option we might consider is to use `Unix.fstat` but I assume this is a blocking operation (and part of `Unix` which would break `Unix`-less platforms) so that can't be done sensibly?

Reply:
1. Yes. It is a file determined by calling `Fd.infer_using_stat`, and apparently the only platform who cares is the s390x which is a big endian platform. One difference I can spot, is that if the fd is a `socket`, async will try to set the fd in non blocking mode - but not files as linux will ignore that anyways. Why this only has an effect on big endian systems is beyond me.

2. We could call `Unix.fstat` to determine if the socket supports non-blocking, but I don't really think it matters. The type is used mainly to control setting blocking vs non-blocking options on the file descriptor. We never reads or writes to the file descriptor, but just read the state (if reads or writes _would_ block), so setting non_blocking is a noop seen from our use case.  But... The zmq documentation indicates that the file descriptor returned is used internally also so modifying it seems a bad idea. 

Digging further into this, `Fd.create` has an option `?avoid_nonblock_if_possible:bool` (`async_unix.0.14`) (and `?avoid_setting_nonblock:bool` for `async_unix.0.15`). Setting this to true keeping the kind as `Fd.Kind.Socket 'Active`, also fixes the problem. When not set, the test spins using 100% cpu forever.

If only janestreet would keep the API stable, we could have forced never setting the non-blocking mode. For now we need to rely on the fact that non-blocking is never set on files. 

In conclusion, I don't think we need to care too much about the actual type, but just make sure that we never change the fd to non blocking.